### ISSUE TITLE
Spurious VUID-vkAcquireNextImageKHR-surface-07783 Error

### ISSUE THREAD

**Issue by greg-churchillnavigation** [KhronosGroup/Vulkan-ValidationLayers#8753]
**Environment:**
 - OS: Windows 11 Pro build 22631.4317
 - GPU: nVidia RTX 2000 driver v31.0.15.5274
 - SDK: 1.3.296.0
 - Options enabled: validation, synchronization:

**Describe the Issue**

App has 2x OS windows and swapchains
When one window is minimized, vkQueuePresentKHR() returns VK_ERROR_OUT_OF_DATE_KHR.
App sees swapchain current surface extent 0,0, so skips rendering that window.
Later, window is restored, (as code was skipping zero size window, nothing was done)
swapchain vkAcquireNextImageKHR() returns VK_ERROR_OUT_OF_DATE_KHR, so re-creates swapchain.
	
Logging my calls shows all Acquires have a matching Presents.

Validation error:
```
Validation Error: [ VUID-vkAcquireNextImageKHR-surface-07783 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_SWAPCHAIN_KHR; | MessageID = 0xad0e15f6 | vkAcquireNextImageKHR():  Application has already previously acquired 3 images from swapchain. Only 3 are available to be acquired using a timeout of UINT64_MAX (given the swapchain has 4, and VkSurfaceCapabilitiesKHR::minImageCount is 2).
The Vulkan spec states: If forward progress cannot be guaranteed for the surface used to create the swapchain member of pAcquireInfo, timeout must not be UINT64_MAX (https://vulkan.lunarg.com/doc/view/1.3.296.0/windows/1.3-extensions/vkspec.html#VUID-vkAcquireNextImageKHR-surface-07783)
```
Occurs at vkAcquireNextImageKHR() for OTHER window swapchain, not the one that was minimized and restored.
Despite validation error, vkAcquireNextImageKHR() returns VK_SUCCESS.
	
Work around:
	In call to vkQueuePresentKHR()
	Pass not null array via VkPresentInfoKHR .pResults, to accept per swapchain image present results.
That's all! Don't actually do anything with the returned results.

Does not occur with 
GPU: Intel Graphics driver 32.0.101.6078

**Expected behavior**

Don't expect a validation error. No apparent leaked acquired images.
Do not expect optionally returned Present results to influence.

**Comment by ziga-lunarg** [comment]
I believe this has just been fixed with #8741

**Comment by greg-churchillnavigation** [comment]
No, it doesn't use vkReleaseSwapchainImagesEXT.

(I did attempt to use that as work around for this issue, because the code would sometimes prefer to abort a render phase when swapchain become invalid, due to resize or such. I encountered either the same, or another issue and gave up trying it. It was also unsupported on intel GPU which I was trying to compare behavior with, while isolating.  Instead, it just either destroys & re-creates the swapchain or potentially passes the expired swapchain + image index to the next present, I believe.)

**Comment by ziga-lunarg** [comment]
Then #8741 does not address this issue.

**Comment by artem-lunarg** [comment]
@greg-churchillnavigation some questions to understand this better

> Logging my calls shows all Acquires have a matching Presents.

Did any if these Acquire calls (that have matching present) return vulkan error (VkResult, not validation)?
And similar, did any if there Present calls (that have matching Acquire) return vulkan error?

> Occurs at vkAcquireNextImageKHR() for OTHER window swapchain, not the one that was minimized and restored.

Were there any vulkan errors or validation errors for this OTHER swapchain before the error reported in this issue?

Does minimizing one swapchain is required to get `vkAcquireNextImageKHR` for another swapchain?

**Comment by artem-lunarg** [comment]
What is the general scheme of image acquire and present synchronization? It would be helpful to know if image acquire uses semaphore or fence, how those synchronization primitives are used further in the frame, which semaphore is used in QueuePresent call?

**Comment by artem-lunarg** [comment]
One more question!

> Pass not null array via VkPresentInfoKHR .pResults, to accept per swapchain image present results.

Does the application use a single QueuePresent call to present to two swapchains at the same time?

**Comment by greg-churchillnavigation** [comment]
> One more question!
> 
> > Pass not null array via VkPresentInfoKHR .pResults, to accept per swapchain image present results.
> 
> Does the application use a single QueuePresent call to present to two swapchains at the same time?

Yes, single present, 2x or 3x swap chains, one for each OS window.

**Comment by greg-churchillnavigation** [comment]
> @greg-churchillnavigation some questions to understand this better
> 
> > Logging my calls shows all Acquires have a matching Presents.
> 
> Did any if these Acquire calls (that have matching present) return vulkan error (VkResult, not validation)? And similar, did any if there Present calls (that have matching Acquire) return vulkan error?
> 
> > Occurs at vkAcquireNextImageKHR() for OTHER window swapchain, not the one that was minimized and restored.
> 
> Were there any vulkan errors or validation errors for this OTHER swapchain before the error reported in this issue?
> 
> Is minimization of one swapchain required to get `vkAcquireNextImageKHR` validation error for another swapchain? (basically what I want to know if you have this situation that you perform some manipulations with the first swapchain and do not touch the second window at all, and then it starts reporting errors for the swapchain from the second window).

The OTHER swapchain, which is the OS window not being minimized does not return any error results at any time.

The swapchain being minimized returns an error on QeueuPresent, when minimize first occurs.  Because we stop rendering at this time due to surface capabilities returning zero size surface, the next error occurs on AcquireNextImage, when image is restored.  In both cases, the result is VK_ERROR_OUT_OF_DATE_KHR, I believe.

I recall, after re-creating the restored window swapchain, I get the validation error from AcquireNextImage call from the OTHER window, though this call returns success.

**Comment by greg-churchillnavigation** [comment]
> What is the general scheme of image acquire and present synchronization? It would be helpful to know if image acquire uses semaphore or fence, how those synchronization primitives are used further in the frame, which semaphore is used in QueuePresent call?

Okay, lets try to explain this...

Swapchain helper Class stores 
```
	std::vector<Image*> m_images; // NOTE: Images are owned by swapchain, not this class.
	std::vector<ImageView*> m_imageViews;

	// NOTE: Manage swap chain & present related semaphores here so they can be re-created along with swap chain.
	//       If semaphore was not signaled or not waited upon, its status can become ambiguous and thus impossible to recycle.
	uint32_t m_semaphoreIndex = 0; // Current ring index (not the same as image index)
	// Ring of semaphores to use and recycle. Deleted when reset as status cannot be determined.
	std::vector<VkSemaphore> m_imageSemaphoresRing;
	std::vector<VkSemaphore> m_renderFinishedSemaphoresRing;
	
	// Semaphores ordered by image index, may be null
	std::vector<VkSemaphore> m_imageAcquiredSemaphores;
	std::vector<VkSemaphore> m_renderFinishedSemaphores;
```
Swapchain Re/create function
1) creates new swap chain, passing old if present before destroying old
2) gets all images via vkGetSwapchainImagesKHR
3) transitions layouts to VK_IMAGE_LAYOUT_PRESENT_SRC_KHR and creates views for images
4) creates an array of Acquire semaphores and array of RenderFinished semaphores and sizes matching arrays for ordering.

Swapchain Acquire function
1) checks for lost surface
2) calls vkAcquireNextImageKHR with ring index for acquire semaphore and advances ring
3) if acquire fails in any way, re-creates the swap chain & resources, but returns a custom not-ready state, so caller will not render that frame

The app main thread attempts to render, first attempting to acquire next image in a BeginRender phase.
If that succeeds, rendering occurs, but is actually buffered and Vulkan command buffers are built on a Render thread.
When all rendering for frame completes, Present is also instigated which runs on the render thread.
Basically off loading work from main thread to emulate behavior like D3D11, which is one of our legacy render APIs.

If Acquire (our BeingRender) succeeded, our swapchain ImageAcuired semaphore is buffered in our Device (helper) to queue submit with the next graphics submit as a wait semaphore.
At our EndRender phase, our Device buffers some PresentInfo which includes the RenderFinished semaphore.
When the app frame is done rendering, a queue submit is called with RenderFinished added to signal semaphores. 
(All OS windows have finished rendering for that 'frame' and so swapchains and semaphores are collected together.)
QueuePresent is then called with RenderFinished added to wait semaphores, along with swapchain handles and image indices.
(All queue submits and presents instigated by main thread are buffered and executed on the Render thread.
Other CPU threads can build command buffers etc. Behavior is similar to Han's/Maister's Granite engine.)
The final queue submit for each frame includes a timeline semaphore, which we wait on before re-using that frame.
We no longer use any fences at all, only timeline semaphores for fence purposes.

Because this 'end frame' combination of queue submit and queue present occurs on a Render thread, on occasion, multiple CPU render frames are buffered.
This means multiple images can be acquired at the same time.
If the app/render code needs to DeviceWaitIdle, we first sync the render thread to the main thread and pause render activity.
The swap chain is re-created after a device wait idle, which is about the only time this occurs.

**Comment by artem-lunarg** [comment]
@greg-churchillnavigation thanks a lot for the details. I'll try to parse this and to reproduce.

**Comment by artem-lunarg** [comment]
> Work around:
In call to vkQueuePresentKHR()
Pass not null array via VkPresentInfoKHR .pResults, to accept per swapchain image present results.

That's actually explains what's going on, can see the issue in the code.

**Comment by artem-lunarg** [comment]
@greg-churchillnavigation this should be fixed in the latest code. I managed to reproduce the same behavior (and that pResults is a workaround) and now it works without the need to specify pResults.
